ccmlib/common.py

```diff
@@ -65,6 +66,118 @@ class UnavailableSocketError(CCMError):
     pass
 
 
+class LockFile:
+    """
+    A class to create filesystem-based lockfiles that are not left
+    locked when process exits unexpectedly.
+    
+    This class uses locking mechanisms provided by kernel. Currently
+    that's fcntl.flock, due to its more intuitive semantics, but if
+    compatibility becomes a problem, it should be possible to switch to
+    fcntl.lockf or use platform-specific functions.
+
+    Lockfile stores PID of owning process and a "status" - which is an
+    arbitrary string that owning process can set. Format of the file is:
+    PID;status
+
+    This class is not thread safe.
+
+    Currently it is only used in ccm to prevent concurrent downloads,
+    so logging messages in __enter__ are specific to this use case.
+    If this class is ever needed somewhere else, this can be changed,
+    either by changing messages or the API.
+
+    Attributes
+    ----------
+    filename: str | bytes | os.PathLike
+        Path to a lockfile - used for logging
+    file: TextIO
+        File handle which will be used to take a lock.
+    locked: bool
+        Current status of the lock.
+    """
+    filename: str | bytes | os.PathLike
+    file: TextIO
+    locked: bool
+
+    def __init__(self, filename: str | bytes | os.PathLike):
+        # We use append because:
+        # - if a file doesn't exist, we need to create it
+        # - we don't want to truncate existing file
+        # - we want RW access to file
+        # 'a+' is the only mode that satisfies all of this.
+        self.filename = filename
+        self.file = open(filename, 'a+')
+        self.locked = False
+
+    def try_lock(self, new_status: str = None) -> (bool, Optional[int]):
+        """Tries to take a lock. If it fails, it return PID of the process
+        that currently owns this lock.
+        """
+        assert not self.locked
```

Comment:
i am confused. if the lock is already owned by some process, `self.locked` is True. so in this case, `(False, blocking_pid)` should be returned. but in your implementation, `AssertionError` is raised instead.

Reply:
`self.locked` is True if it is locked by this process.

Reply:
This assert will throw if a user tries to take already owned lock, so e.g. a code like this:
```
l = LockFile('lockfile')
(success, pid) = l.try_lock()
assert success # Let's assume there are no other instances
(success, pid) = l.try_lock() # This would throw
```
I can add a comment that the lock is not reentrant.

Reply:
> I can add a comment that the lock is not reentrant.

that'd be helpful. i see. thank you for the detailed explanation. this makes more sense now.

Reply:
Added the note with an example. Also changed description of the `_locked` member.